go/ephemeral/common.go

```diff
@@ -44,3 +46,45 @@ func getCurrentUserUV(ctx context.Context, g *libkb.GlobalContext) (ret keybase1
 	})
 	return ret, err
 }
+
+// Map generations to their creation time
+type keyExpiryMap map[keybase1.EkGeneration]keybase1.Time
+
+// Keys expire after `KeyLifetimeSecs` unless there has been a gap in their
+// generation. If there has been a gap, a key can be re-used for up to
+// `KeyLifetimeSecs` until it is considered stale. To determine expiration, we
+// look at all of the current keys and account for any gaps since we don't want
+// to expire a key if it is still used to encrypt a different key. This only
+// applies to deviceEKs or userEKs since they can have a dependency above them.
+// A teamEK expires after `KeyLifetimeSecs` without exception, so it doesn't
+// call this.
+func getExpiredGenerations(keyMap keyExpiryMap, nowCTime keybase1.Time) (expired []keybase1.EkGeneration) {
+
+	// Sort the generations we have so we can walk through them in order.
+	var keys []keybase1.EkGeneration
+	for k := range keyMap {
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
+
+	var nextCTime keybase1.Time
+	var expiryOffset keybase1.Time
+	for i, generation := range keys {
+		currentCTime := keyMap[generation]
+		if i < len(keys)-1 {
+			nextCTime = keyMap[keys[i+1]]
+			expiryOffset = nextCTime - currentCTime
+			if expiryOffset > KeyLifetimeSecs { // Offset can be max KeyLifetimeSecs
+				expiryOffset = KeyLifetimeSecs
+			}
+		} else {
+			expiryOffset = 0
```

Comment:
I think this is saying that the expiry offset for the latest key is always zero. But that doesn't seem right. If the latest key is 9 days old, we still can't delete it, because something might've been encrypted for it 3 days ago when it was still valid. I think the latest key should behave like `nextCTime = nowCTime`.

Reply:
i don't think we'd hit this since we generate and then cleanup but agreed it's worthwhile to add